crates/pico/src/memo.rs

```diff
@@ -0,0 +1,188 @@
+use pico_core::{
+    container::Container,
+    database::Database,
+    dyn_eq::DynEq,
+    epoch::Epoch,
+    key::Key,
+    node::{Dependency, DerivedNode, DerivedNodeId, NodeKind},
+    params::ParamId,
+    storage::{Storage, StorageMut},
+};
+
+pub enum MemoState {
+    Memoized,
+    Computed,
+}
+
+#[allow(clippy::map_entry)]
+pub fn memo<Db: Database>(
+    db: &mut Db,
+    node_id: DerivedNodeId,
+    inner_fn: fn(&mut Db, ParamId) -> Box<dyn DynEq>,
+) -> MemoState {
+    let mut state = MemoState::Memoized;
+    let current_epoch = db.current_epoch();
+    let time_calculated = if db.storage().derived_nodes().contains_key(&node_id) {
+        if any_dependency_changed(db, &node_id, current_epoch) {
```

Comment:
Can we restructure these lines to be like

```
match db.storage().derived_nodes().get(&node_id) {
  Some(x) => if any_dependency_changed(db, x, current_epoch)
}
```

So that we can avoid the `expect` in `any_dependency_changed`?

Reply:
`db` is borrowed as mutable in `any_dependency_changed` but we've already borrowed it in `db.storage().derived_nodes().get(&node_id)`